### PR TITLE
upgrade clap to v4.5

### PR DESCRIPTION
**Pull request by eserilev**
## Issue Addressed

#5087

## Proposed Changes

This PR upgrades clap from version 2.x to 4.5.1. There are some breaking changes between the two versions, v4.x is a bit stricter when it comes to flag definitions and has a slightly different api. Once we upgrade clap to v4 we can begin working on migrating to clap derive, which gives us the following benefits

- easier to read cli flag definitions
- a bit more type safety/compile time guarantees
- less verbose initialization logic
-  an easy interface to enable reading cli flags from a config file

We'll migrate to clap derive in separate PR's on a crate by crate basis, to make the PRs a bit more manageable

### List of breaking changes:

- `latency-measurement-service` flag is now deprecated in favor of `disable-latency-measurement-service`. By default the latency measurement service is enabled, and can be disabled by providing the new `disable-latency-measurement-service` flag.
- The `self-limiter` flag no longer accepts arguments. Providing the `self-limiter` flag enables the outbound rate limiter service. A new flag `self-limiter-protocols` is introduced to allow for configuring the outbound rate limiter service per protocol. In order to use the new `self-limiter-protocols` flag, `self-limiter` must also be provided. 
-  The `inbound-rate-limiter` flag is now deprecated in favor of `disable-inbound-rate-limiter`.  By default the inbound rate limiter service is enabled and can be disabled by providing the new `disable-inbound-rate-limiter` flag. Also, an additional flag `inbound-rate-limiter-protocols` is introduced to allow for configuring the inbound rate limiter service per protocol. Note that the `disable-inbound-rate-limiter` flag conflicts with the `inbound-rate-limiter-protocols` flag, they cannot be used in tandem.
- `slasher-broadcast` now requires either a true or false value. The flag cannot be used without providing a value.
- The above should cover the big breaking changes introduced by this PR. However its important to note that cli flag definitions are a bit stricter now. Anything listed under Options in the help screen must have an argument provided. Anything listed under Flags in the help screen cannot have an argument provided. If those conditions are violated the program will fail and an error message will be returned specifying what cli rules were violated.

### CODE REVIEW

**Comment by eserilev** [comment]
There are still roughly 10 failing tests here that I need to resolve

File: validator_client/src/cli.rs
Comment (eserilev) [inline validator_client/src/cli.rs]: Introduced a breaking change here

`latency-measurement-service` is currently accepting "true" or "false" as values. If the `latency-measurement-service` flag is included without a value, it'll be treated as `true`. This causes issues with the default `ArgAction` behaviors intoduced in clap v4.

We have a few `ArgAction` options here, the two most relevant being `ArgAction::Set` and `ArgAction::SetTrue` (or `SetFalse)

`ArgAction::Set` requires a value to be set, it'll panic if no value is set
`ArgAction::SetTrue` requires that no value be set, it'll panic if a value is set

`latency-measurement-service` is currently set up as a hybrid between `Set` and `SetTrue`. There is a workaround in clap v4 to enable this behavior

```
.default_missing_value("true")
.require_equals(true)
.num_args(0..1)
```

`default_missing_value("true")` requires `num_args(0..n)` AND `required_equals(true)` to unambiguously determine what, if any, value was supplied for the argument.  We could set `required_equals` to false, but it could cause some unintended side effects like in the example below:

`--latency-measurement-service subcommand --other-flag`

subcommand would actually get treated as the value for `latency-measurement-service` flag. With `requires_equals(true)` this issue could never happen, since an equals sign wasn't supplied. In practice this situation almost probably never happens. But it still feels "bad" 

My recommendation here is to deprecate `latency-measurement-service` and introduce a new flag `disable-latency-measurement-service` with `ArgAction::SetFalse`. This means by default the latency measurement service is enabled, unless specifically disabled with the flag.

File: beacon_node/src/cli.rs
Comment (eserilev) [inline beacon_node/src/cli.rs]: see comment regarding `--latency-measurement-service` flag why this configuration feels "bad".

I suggest introducing a new flag `--enable-self-limiter` with `ArgAction::SetTrue` and changing the `--self-limiter` flag to `ArgAction::Append`.

if the `self-limiter` flag is provided with value(s), self-limiter will be enabled with the provided value(s).  if `--enable-self-limiter` flag is provided, the quotas for the inbound rate limiter will be used. (Also do the same for the `inbound-rate-limiter` flag)

**Comment by eserilev** [comment]
Tests are passing, this should be ready for a review

File: validator_client/src/cli.rs
Comment (jxs) [reply validator_client/src/cli.rs]: Hi Eitan, we could also keep it as a string to maintain the same behaviour, refraining from levegeraring the type system just for this one right? If so we can soft deprecate introducing the new one and then hard deprecate on the next major release

File: beacon_node/src/cli.rs
Comment (AgeManning) [inline beacon_node/src/cli.rs]: Can probably remove this comment. There isn't a way to enforce 2 values? I guess we can check at validation time.

File: beacon_node/src/cli.rs
Comment (AgeManning) [inline beacon_node/src/cli.rs]: Same here

File: beacon_node/src/cli.rs
Comment (AgeManning) [reply beacon_node/src/cli.rs]: Yep, lets do this to make it more compatible with the new clap

File: beacon_node/src/config.rs
Comment (AgeManning) [inline beacon_node/src/config.rs]: We can remove this old code now, right?

File: boot_node/src/cli.rs
Comment (AgeManning) [inline boot_node/src/cli.rs]: and here

File: boot_node/src/cli.rs
Comment (AgeManning) [inline boot_node/src/cli.rs]: and here

File: testing/simulator/src/cli.rs
Comment (AgeManning) [inline testing/simulator/src/cli.rs]: can't do version anymore?

**Review by AgeManning — COMMENTED** [review]
I had a quick skim. Looks good to me! 

I'm of the opinion that we make all the breaking changes we need right now. Adjust all the CLI flags we need, regardless if they break. 

I'll try and do a follow up PR which tries to hide some flags and remove deprecated flags, then in the next release we can warn users of breaking CLI changes.

File: account_manager/src/wallet/create.rs
Comment (dapplion) [inline account_manager/src/wallet/create.rs]: Why declare as separate function? `value_parser` can take a closure, as long as you annotate the type:

```rust
                .value_parser(|len: &str| {
                    match len.parse::<usize>().ok().and_then(|words| MnemonicType::for_word_count(words).ok()) {
                        Some(_) => Ok(len.to_string()),
                        None => Err(format!(
```

File: beacon_node/src/cli.rs
Comment (dapplion) [reply beacon_node/src/cli.rs]: You have to use
```
                .num_args(0..=2)
```

File: boot_node/src/cli.rs
Comment (dapplion) [inline boot_node/src/cli.rs]: Why are these options being added for the bootnode? It shouldn't care about validating bn flags like `subscribe-all-subnets`, `light-client-server`, `proposer-only`

File: common/clap_utils/src/lib.rs
Comment (dapplion) [inline common/clap_utils/src/lib.rs:203]: Will this render all text as green?

File: lighthouse/src/main.rs
Comment (dapplion) [inline lighthouse/src/main.rs:26]: Why move into static refs?

EDIT: Now `.long_version` expects as 'static str

File: testing/simulator/src/cli.rs
Comment (dapplion) [reply testing/simulator/src/cli.rs]: It's possible, you need to add the feature
```yaml
clap = { version = "4.5.1", features = ["cargo"] }
```
and import
```rust
use clap::{crate_version, Arg, ArgAction, Command};
```

File: validator_manager/src/create_validators.rs
Comment (dapplion) [inline validator_manager/src/create_validators.rs]: Should the validator manager be able to dump-config?

File: common/clap_utils/src/lib.rs
Comment (eserilev) [reply common/clap_utils/src/lib.rs:203]: no, it looks like this:

image

**Comment by chong-he** [comment]
I have done some testing on this and some comments below (`current` refers to the help texts that we have now, while `new` refers to the help texts from this PR):  

1.	The help texts work on both Linux and Windows. The help texts are displaying for all commands and subcommands, with green colour on the flag itself, and white colour for the description. Only some current help texts (such as bn and vc) have this colour distinction, while other commands like ` lighthouse a validator –help` and `lighthouse a wallet –help` do not. So this is a great improvement. 

2.	The current help texts always generate the commit in the first line, in which @macladson removed it in the `cli.sh` when putting them in the Lighthouse Book, as if otherwise the `cli-check` would always fail. The new help texts do not have this commit in the first line anymore, and as such the removal of first line is no longer needed. If we run `make cli-local` now, the help texts will always start with this:

`
Usage: lighthouse beacon_node [OPTIONS]
`
      
which is not desirable as the important descriptive first line is removed. To solve this, we remove these lines:

https://github.com/sigp/lighthouse/blob/3058b96f2560f1da04ada4f9d8ba8e5651794ff6/scripts/cli.sh#L15-L16

I have tested that after removing the above in `cli.sh` and running `make cli-local`, all help texts are reserved and shown.

3. Currently, the flags are separated into 2 groups: flags (do not require a value), options (require a value). With this PR, all flags are listed under options only. Is this an expected change? 

4. Currently the flags are arranged in alphabetical order. I am not sure in what order is the new help texts arranged? For example, in `lighthouse bn –help`, it starts with  the flags `--network-dir`, then `--freezer-dir` and `--logfile` which does not seem follow a certain order? It would be nice if they follow a certain order, e.g., the alphabetical order. 

5. The new help texts do not start a new line until reaching the monitor width. This unfortunately breaks the word when starting a new line, referring to the screenshot below. The words `exceeds` and `attestations` are broken into two lines. 

newline

The current help texts do not have this problem because it starts a new line after a certain number of words. Because the new help texts is shown as a single line, when it is displayed in Lighthouse book, it will have a long horizontal scroll at the bottom, which is not very UX friendly:

scroll

I wonder if it is a good idea to break into new lines after reaching a certain number of words, like the current help texts?

Thanks!

File: beacon_node/src/config.rs
Comment (dapplion) [inline beacon_node/src/config.rs:984]: If dropped

File: beacon_node/src/config.rs
Comment (dapplion) [inline beacon_node/src/config.rs:1438]: Changing from `inbound-rate-limiter` to `disable-inbound-rate-limiter` appears to be a breaking change on the beacon subcommand, is it necessary?

AFAIU the current flag usage is to disable is `--inbound-rate-limiter=disable`?

Comment (eserilev) [reply beacon_node/src/config.rs:1438]: Currently this cli param is "bad" because it tries to do too many things. By default its value is set to `true`. if the cli param is provided without a value, its treated as `true`. On top of this the param can either accept the value `disable` OR a `;` separated list of protocol values w/ the following format: `<protocol_name>:<tokens>/<time_in_seconds>`

Technically these breaking changes aren't necessary, however the new version of clap is a bit stricter with cli param definitions. We now either define the cli param as accepting a value (`ArgAction::Set`) or as a bool flag (`ArgAction::SetTrue`). The existing functionality requires us to treat this cli param as a hybrid of `Set` and `SetTrue` like so

```
.default_missing_value("true")
.require_equals(true)
.num_args(0..1)
```

which is also technically a breaking change since the cli param now requires an `=` sign if the user tries to provide it a value (currently the `=` isn't necessary). I think its a bit nicer to just keep all cli params as either `Set` or `SetTrue`, but I'm open to the hybrid option if we decide this breaking change isnt worth it.

**Comment by eserilev** [comment]
@chong-he thanks for the detailed review! I think i was able to fix the issues you brought up

- the cli params are now listed in alphabetic order
- I created a separate "Flag" section for cli params that dont accept a value
- text wrapping should be a bit cleaner
- i updated cli.sh based on your recommendation

File: beacon_node/src/config.rs
Comment (eserilev) [reply beacon_node/src/config.rs:984]: thanks, resolved in https://github.com/sigp/lighthouse/pull/5273/commits/4555fb27362cc22166730d6049977f208b4e9a3e

**Comment by eserilev** [comment]
In an attempt to keep all flags set to either `ArgAction::Set`, `ArgAction::Append`, or `ArgAction::SetTrue` I've introduced a few breaking changes

### Self Limiter

the `self-limiter` flag no longer takes a value. If the flag is provided it'll enable the outbound rate limiter. I've also introduced a second flag `self-limiter-protocols` that takes a list of protocol values (delimited by a `;`). If `self-limiter` is provided, w/o `self-limiter-protocols`, the outbound rate limiter will be enabled with the default values


### Inbound rate limiter

I've deprecated the `inbound-rate-limiter` flag and instead introduced a `disable-inbound-rate-limiter` flag. By default inbound rate limiting is enabled, and can only be disabled via this new flag. Additionally I've introduced a `inbound-rate-limiter-protocols` flag that accepts a list of protocol values (delimited by a `;`. 

We could make these breaking changes as a separate PR instead, though I'd rather introduce these breaking changes before merging this in.

**Comment by chong-he** [comment]
> @chong-he thanks for the detailed review! I think i was able to fix the issues you brought up
> 
> * the cli params are now listed in alphabetic order
> * I created a separate "Flag" section for cli params that dont accept a value
> * text wrapping should be a bit cleaner
> * i updated cli.sh based on your recommendation

Looks great. I can confirm that the flags are now arranged in alphabetical order, and the flags are separated into `flags` and `options`. 

I notice for some commands, `options` flags, the description is now side-by-side with the flag, instead of the description below the flag. For example, `lighthouse vm move --help`:

vm

(bn and vc are still having the description below the flag)

I see that the `cli-check` failed. To fix this, you can run `make cli-local` then commit the changes (it will auto update all `help_x.md` files)

File: account_manager/src/validator/exit.rs
Comment (chong-he) [inline account_manager/src/validator/exit.rs:36]: I think there is a bug here. When I test with:
`lighthouse account validator exit --keystore staking/keystore-m_12381_3600_7_0_0-1704891376.json`

It returns:
```
 CRIT Task panic. This is a bug!              advice: Please check above for a backtrace and notify the developers, backtrace:    0: lighthouse::run::{{closure}}
   1: <alloc::boxed::Box<F,A> as core::ops::function::Fn<Args>>::call
             at /rustc/07dca489ac2d933c78d3c5158e3f43beefeb02ce/library/alloc/src/boxed.rs:2029:9
   2: std::panicking::rust_panic_with_hook
             at /rustc/07dca489ac2d933c78d3c5158e3f43beefeb02ce/library/std/src/panicking.rs:783:13
   3: std::panicking::begin_panic_handler::{{closure}}
             at /rustc/07dca489ac2d933c78d3c5158e3f43beefeb02ce/library/std/src/panicking.rs:657:13
   4: std::sys_common::backtrace::__rust_end_short_backtrace
             at /rustc/07dca489ac2d933c78d3c5158e3f43beefeb02ce/library/std/src/sys_common/backtrace.rs:171:18
   5: rust_begin_unwind
             at /rustc/07dca489ac2d933c78d3c5158e3f43beefeb02ce/library/std/src/panicking.rs:645:5
   6: core::panicking::panic_fmt
             at /rustc/07dca489ac2d933c78d3c5158e3f43beefeb02ce/library/core/src/panicking.rs:72:14
   7: clap_builder::parser::matches::arg_matches::ArgMatches::get_flag
   8: account_manager::validator::exit::cli_run
   9: account_manager::validator::cli_run
  10: account_manager::run
  11: lighthouse::main::{{closure}}
  12: lighthouse::main
  13: std::sys_common::backtrace::__rust_begin_short_backtrace
  14: std::rt::lang_start::{{closure}}
  15: core::ops::function::impls::<impl core::ops::function::FnOnce<A> for &F>::call_once
             at /rustc/07dca489ac2d933c78d3c5158e3f43beefeb02ce/library/core/src/ops/function.rs:284:13
  16: std::panicking::try::do_call
             at /rustc/07dca489ac2d933c78d3c5158e3f43beefeb02ce/library/std/src/panicking.rs:552:40
  17: std::panicking::try
             at /rustc/07dca489ac2d933c78d3c5158e3f43beefeb02ce/library/std/src/panicking.rs:516:19
  18: std::panic::catch_unwind
             at /rustc/07dca489ac2d933c78d3c5158e3f43beefeb02ce/library/std/src/panic.rs:142:14
  19: std::rt::lang_start_internal::{{closure}}
             at /rustc/07dca489ac2d933c78d3c5158e3f43beefeb02ce/library/std/src/rt.rs:148:48
  20: std::panicking::try::do_call
             at /rustc/07dca489ac2d933c78d3c5158e3f43beefeb02ce/library/std/src/panicking.rs:552:40
  21: std::panicking::try
             at /rustc/07dca489ac2d933c78d3c5158e3f43beefeb02ce/library/std/src/panicking.rs:516:19
  22: std::panic::catch_unwind
             at /rustc/07dca489ac2d933c78d3c5158e3f43beefeb02ce/library/std/src/panic.rs:142:14
  23: std::rt::lang_start_internal
             at /rustc/07dca489ac2d933c78d3c5158e3f43beefeb02ce/library/std/src/rt.rs:148:20
  24: main
  25: __libc_start_call_main
             at ./csu/../sysdeps/nptl/libc_start_call_main.h:58:16
  26: __libc_start_main_impl
             at ./csu/../csu/libc-start.c:392:3
  27: _start
, message: arg `no-wait`'s `ArgAction` should be one of `SetTrue`, `SetFalse` which should provide a default, location: /home/hi/.cargo/registry/src/index.crates.io-6f17d22bba15001f/clap_builder-4.5.1/src/parser/matches/arg_matches.rs:181:17
``` 

All other `lighthouse account validator` and `lighthouse account wallet` flags are tested and they are good, just this one.

File: beacon_node/src/config.rs
Comment (dapplion) [reply beacon_node/src/config.rs:1438]: Makes sense. If @AgeManning blesses this breaking change I'm okay with it

**Comment by dapplion** [comment]
Do we really want to split flags and options? I never understand the UX benefit of doing so. IMO flags should either:
- be sorted in alphabetical order (flags and options together)
- be grouped in some way (network, log, mev) and then sorted in alphabetical order (flags and options together)

**Comment by eserilev** [comment]
I really like the idea of grouping the flags as Lion describes (network, log, mev etc.)

**Comment by chong-he** [comment]
> Do we really want to split flags and options? I never understand the UX benefit of doing so. IMO flags should either:
> 
> * be sorted in alphabetical order (flags and options together)
> * be grouped in some way (network, log, mev) and then sorted in alphabetical order (flags and options together)

Makes sense, I was commenting by comparing to previous help texts arrangement. But grouping them does sound good because if it doesn't require a value, it will be clear by looking at the flag itself

File: account_manager/src/validator/exit.rs
Comment (eserilev) [reply account_manager/src/validator/exit.rs:36]: this one should be fixed

Comment (chong-he) [reply account_manager/src/validator/exit.rs:36]: yeah tested and it is good now, thanks

File: beacon_node/src/cli.rs
Comment (chong-he) [inline beacon_node/src/cli.rs:80]:
```suggestion
                .help_heading(FLAG_HEADER)
                .display_order(0)
```

File: beacon_node/src/cli.rs
Comment (chong-he) [inline beacon_node/src/cli.rs]: Not sure why the short name here is `x`? This flag appears just before `zero-ports` which is seperated from other `--disable` flags. Current flags although with x, but it still shows in alphabetical order starting with `d`

**Review by chong-he — COMMENTED** [review]
Some alphabetical order comments

File: beacon_node/src/cli.rs
Comment (eserilev) [reply beacon_node/src/cli.rs]: i ended up removing the short name

**Comment by eserilev** [comment]
re: cli param grouping

in subsequent PRs im going to move us to clap derive on a crate by crate basis. At that time I think I'll look to adding more granular/logical cli param grouping. For now I've kept it as is (i.e. flag vs option grouping). If we're against this, I can do no groupings and just have it purely alphabetical. 

aside from that, I think everything else here seems to be in a good spot. thank you @chong-he for all the help!

**Comment by realbigsean** [comment]
confused why the cli-check is failing. It's working for me locally 🤔

**Comment by chong-he** [comment]
> Member



> confused why the cli-check is failing. It's working for me locally 🤔

I will look into this

**Comment by chong-he** [comment]
Some comments after taking another look:

1.	Should `lighthouse boot_node --enr-address` take a value? For now, with `lighthouse boot_node --enr-address` it doesn’t return an error.

2.	`lighthouse boot_node --network-dir dummy_directory` returns `9000` when a dummy directory is set. Not sure why the 9000 is returned?

3.	Should the subcommands follow alphabetical order too? Currently the subcommands, e.g., in `lighthouse –help`, the commands `database_manager`, `validator_manager` etc do not follow the alphabetical order. 

4.	If we still separate `options` and `flags`, the `--help` should be in `flags` since it doesn’t take a value. They are in `options` now. 

5.	I find that `--testnet-dir` is appearing first, although if following alphabetical order, it should come after other flags, such as `--target-peers` in bn,  `--terminal-xxx` in vc. It seems that it has something to do with the fact that ` .short('t')` is put in the `--testnet-dir` flag?

6.	Do we want to consider this? https://github.com/sigp/lighthouse/pull/5273#issuecomment-2038467089, i.e., having the descriptions below the flags. Currently only bn and vc are having the description below. 

    Are we also open to having more text wrapping? This is mainly so that the help texts are displayed on the Lighthouse book without a horizontal scroll like what we currently have: https://lighthouse-book.sigmaprime.io/help_vm.html

    The new one will have a horizontal scroll which is not very UX friendly.
vm

**Comment by eserilev** [comment]
thanks @chong-he 
1. fixed, it should now return an error when a value isnt provided
2. the random "9000" has been removed
3. subcommands now display in alphabetical order
4. help should now show up under the flags section
5. yes, the alphabetical order takes into account the short name `-t`. i dont see a way to override this functionality
6. i've moved descriptions to below the flag everywhere. let me know what you think
7. ive added better text wrapping, i think the book looks a bit better now (see screenshot below)

image

**Comment by chong-he** [comment]
> thanks @chong-he
> 
> 1. fixed, it should now return an error when a value isnt provided
> 2. the random "9000" has been removed
> 3. subcommands now display in alphabetical order
> 4. help should now show up under the flags section
> 5. yes, the alphabetical order takes into account the short name `-t`. i dont see a way to override this functionality
> 6. i've moved descriptions to below the flag everywhere. let me know what you think
> 7. ive added better text wrapping, i think the book looks a bit better now (see screenshot below)
> 
> image

Looks all good to me, thank you! 

One way to let the `testnet-dir` flag follow the alphabetical order is to remove the `short(t)` I think? But I don't know, do we want to keep the short name there?

**Comment by AgeManning** [comment]
Should we get this in for 5.2?

File: Cargo.lock
Comment (michaelsproul) [inline Cargo.lock:21]: Let's open an issue for upgrading criterion to 0.5 so we can remove the clap 2 dependency here: 

https://github.com/sigp/lighthouse/blob/9a394675dce0a94304c571cf4162c9b474a25fd0/Cargo.lock#L1552

Latest criterion depends on clap 4 as well: https://crates.io/crates/criterion/0.5.1/dependencies

File: lighthouse/tests/beacon_node.rs
Comment (michaelsproul) [inline lighthouse/tests/beacon_node.rs:2137]: breaking change we should flag in the release notes

**Review by michaelsproul — APPROVED** [review]
LGTM.

I would like to see a collected list of breaking changes that we can add to the release notes.

**Comment by eserilev** [comment]
i'll put together a list of breaking changes

File: Cargo.lock
Comment (eserilev) [reply Cargo.lock:21]: will open an issue. theres also this open issue in criterion that the maintainer created to make clap an optional dependency https://github.com/bheisler/criterion.rs/issues/781

File: beacon_node/src/config.rs
Comment (AgeManning) [reply beacon_node/src/config.rs:1438]: Yeah I'm down for breaking changes if it conforms better to the new clap. Makes sense to me.

**Comment by michaelsproul** [comment]
@mergifyio queue

**Pull request merged**
